File: src/shared/Utils/Array.cs
using Lexer;

namespace Utils
{
    public class Array
    {
        // Check whether a string is empty.
        public static bool IsEmpty(string[] arr)
        {
            if (arr.Length == 0 || arr == null) return true;
            else return false;
        }

        // Remove empty strings from an array.
        public static string[] Reduce(string[] arr)
        {
            return arr.Where(x => !String.IsEmpty(x)).ToArray();
        }

        // Remove duplicate elements from an array.
        public static string[] Filter(string[] arr)
        {
            return arr.Distinct().ToArray();
        }

        // Trim the leading part of an array.
        public static string[] TrimStart(string[] arr)
        {
            int Count = 0;
            foreach (string i in arr)
            {
                if (String.IsEmpty(i)) Count++;
                else break;
            }

            return arr.Skip(Count).ToArray();
        }

        // Trim the trailing part of an array.
        public static string[] TrimEnd(string[] arr)
        {
            int Count = 0;
            foreach (string i in arr.ToArray().Reverse())
            {
                if (String.IsEmpty(i)) Count++;
                else break;
            }

            return arr.ToArray().Reverse().Skip(Count).Reverse().ToArray();
        }

        // https://stackoverflow.com/a/70991343/18121288
        // Trim the leading and trailing part of an array.
        public static string[] Trim(string[] arr)
        {
            // No need to search through nothing
            if (arr == null || arr.Length == 0)
                return [];

            // Define predicate to test for non-empty strings
            static bool IsNotEmpty(string s) => !String.IsEmpty(s);

            var FirstIndex = System.Array.FindIndex(arr, IsNotEmpty);

            // Nothing to return if it's all whitespace anyway
            if (FirstIndex < 0)
                return [];

            var LastIndex = System.Array.FindLastIndex(arr, IsNotEmpty);

            // Calculate size of the relevant middle-section from the indices
            var NewArraySize = LastIndex - FirstIndex + 1;

            // Create new array and copy items to it
            var Results = new string[NewArraySize];
            System.Array.Copy(arr, FirstIndex, Results, 0, NewArraySize);

            return Results;
        }
    }
}

File: src/shared/array/array.cpp
#include "aopch.h"
#include "array.h"

#include "strings/strings.h"

namespace array
{
    bool is_empty(const std::vector<std::string>& arr)
    {
        return arr.empty() || arr.size() == 0;
    }

    std::vector<std::string> reduce(const std::vector<std::string>& arr)
    {
        if (is_empty(arr))
            return {};

        std::vector<std::string> result;
        std::copy_if(arr.begin(), arr.end(), std::back_inserter(result), [](const std::string& str) { return !str.empty(); });

        return result;
    }

    // https://stackoverflow.com/a/70991343/18121288
    // https://github.com/SrijanSriv211/AOs/blob/AOs-2.7/src/shared/Utils/Array.cs#L52
    std::vector<std::string> trim(const std::vector<std::string>& arr)
    {
        if (is_empty(arr))
            return {};

        auto is_not_empty = [](const std::string& str) { return !strings::is_empty(str); };

        // find the index of the first non-empty string
        auto first_index = std::find_if(arr.begin(), arr.end(), is_not_empty);

        // nothing to return if it's all whitespace anyway
        if (first_index == arr.end())
            return {};

        // find the index of the last non-empty string
        auto last_index = std::find_if(arr.rbegin(), arr.rend(), is_not_empty).base();

        // create new vector with the trimmed elements
        return std::vector<std::string>(first_index, last_index);
    }
}
